### PR TITLE
perf(api): group the has-record subject ids in the database

### LINKED ISSUE #1413: Performance: 'has records' subject filter reads every instrument record to compute a distinct subject list
### Current Implementation and Limitations

The datahub's "With records only" filter is implemented by reading **every instrument record in scope** into the API process, deduplicating subject ids in JavaScript, and then sending them all back to MongoDB as an `$in` array.

`SubjectsService.querySubjectIdsWithRecords` (`apps/api/src/subjects/subjects.service.ts:161`):

```ts
private async querySubjectIdsWithRecords(groupId?: string): Promise<string[]> {
  const records = await this.prismaClient.instrumentRecord.findMany({
    distinct: ['subjectId'],
    select: { subjectId: true },
    where: groupId ? { groupId } : {}
  });
  return records.map((r) => r.subjectId);
}
```

consumed at `:139`:

```ts
return this.subjectModel.findMany({
  where: {
    AND: [
      accessibleQuery(ability, 'read', 'Subject'),
      groupInput,
      { id: { in: await this.querySubjectIdsWithRecords(groupId) } }
    ]
  }
});
```

Two problems:

1. **`distinct` is not pushed into MongoDB.** Prisma applies `distinct` in the query engine, not as a server-side `$group`/`distinct` command, so the driver receives one document per *record*, not per distinct subject. The deduplication to a much smaller set happens only after all of them have crossed the wire.
2. **The result is round-tripped back as a literal `$in` array.** Every subject id that has any record is serialised into the second query. On a large group this is a query document containing thousands of strings.

Also note `where: groupId ? { groupId } : {}` ignores the caller's `ability` entirely — for the unscoped (root) case this reads the whole `InstrumentRecordModel` collection, regardless of what the requesting user is permitted to see.

This is the issue previously raised as #1346 ("Improve initial hasRecord filter performance", closed). The current implementation still has the shape described above.

### Measurements

MongoDB 8, 200,000 instrument records across 8 groups, 5,000 subjects:

```
scope        docs examined   docs returned to Prisma   distinct subjects   time
group            200,000              25,000                  625          62 ms   (no index)
group             25,000              25,000                  625          42 ms   (with index)
root (all)       200,000             200,000                5,000         ~85 ms   (no index)
```

So for the group case, Prisma transfers **25,000 documents to compute a 625-element result**, then issues a second query whose `where` clause embeds those 625 ids. For the unscoped case it transfers 200,000 documents to compute 5,000 ids. The ratio of transferred documents to useful output is 40:1 and grows with records-per-subject.

Indexing `{groupId, date}` (see the indexing issue) fixes the scan but not the transfer — the 25,000 documents still cross the wire.

### Associated Application Components

Server

### Proposed Solution

**1. Do the distinct in the database.** A `$group` aggregation returns one document per subject instead of one per record:

```ts
private async querySubjectIdsWithRecords(groupId?: string, ability?: AppAbility): Promise<string[]> {
  const result = await this.prismaClient.instrumentRecord.aggregateRaw({
    pipeline: [
      { $match: groupId ? { groupId: { $oid: groupId } } : {} },
      { $group: { _id: '$subjectId' } }
    ]
  });
  return (result as unknown as { _id: string }[]).map((r) => r._id);
}
```

With the `{groupId, date}` index this is a covered-ish scan and returns 625 documents instead of 25,000 — a 40× reduction in data transferred, for the same result.

**2. Better still, avoid the round trip entirely.** Prisma can express "subject has at least one record" as a relation filter, so the two queries collapse into one and no id list is ever materialised in the application:

```ts
return this.subjectModel.findMany({
  where: {
    AND: [
      accessibleQuery(ability, 'read', 'Subject'),
      groupInput,
      { instrumentRecords: { some: groupId ? { groupId } : {} } }
    ]
  }
});
```

This is the version worth trying first — it is simpler than the current code, not more complex. It is worth `explain()`-ing on a realistic dataset before committing, since MongoDB relation filters in Prisma can compile to a `$lookup`, and the aggregation in (1) is the fallback if the plan is worse.

**3. Respect the ability.** `querySubjectIdsWithRecords` should apply `accessibleQuery(ability, 'read', 'InstrumentRecord')` to its own `where`, matching every other query in the service. Currently a request that resolves to the unscoped branch reads records the user has no permission to read (their ids are then intersected away by the outer query, so this is not a data leak, but it is unnecessary work and an inconsistency).

### Estimated Difficulty

Medium

### Priority

Medium

### PR DESCRIPTION
**Pull request by gdevenyi**
Fixes #1413.

The "with records only" datahub filter resolved its subject list with:

```ts
findMany({ distinct: ['subjectId'], select: { subjectId: true }, where })
```

Prisma applies `distinct` in the **query engine**, not in MongoDB, so the driver receives one row per *record* and collapses them to one row per subject only after they have arrived. On a group holding 100k records that is 100k rows transferred to produce a list of at most a few thousand ids.

`groupBy` asks MongoDB for the distinct values directly.

## Measured over HTTP, live instance, 200,300 records, median of 7

| | `main` | this branch |
|---|---|---|
| `GET /v1/subjects?hasRecord=true` | **366 ms** | **42 ms** |
| `GET /v1/subjects?groupId=<g>&hasRecord=true` | 67 ms | 37 ms |
| `GET /v1/subjects?groupId=<g>` (filter off, for reference) | 4 ms | 3 ms |

The unscoped case is where the difference shows, and it widens with the collection: `groupBy` returns one row per subject regardless of how many records back it, so its cost is flat in records-per-subject where the old form was linear.

Confirmed at the query level too — the same comparison run directly through the Prisma client:

```
findMany distinct (current impl), scoped        197ms  -> 55 ids
findMany distinct (current impl), unscoped     1229ms  -> 100 ids
groupBy by subjectId, scoped                     95ms  -> 55 ids
groupBy by subjectId, unscoped                  100ms  -> 100 ids
```

Note `distinct` going 197 → 1229 ms as the scope widens while `groupBy` stays at ~100 ms.

## The issue's "better still" suggestion is wrong — please don't take it

#1413 floated expressing this as a relation filter on Subject:

```ts
{ instrumentRecords: { some: groupId ? { groupId } : {} } }
```

and said it was "the version worth trying first". **I measured it on the same dataset and it takes 26,849 ms** — 270× slower than the code being replaced, let alone the replacement:

```
subject.findMany({instrumentRecords:{some}})   26849ms  -> 55 subjects
```

Prisma compiles a relation filter on MongoDB into a `$lookup` over the whole record collection. That is the same construct that fails outright at scale in #1416 (fixed in #1423 by removing it), and it carries the same 100 MiB per-document ceiling. I've left a comment on the method saying so, and corrected the issue.

## Also fixed: the record query now respects the caller's ability

`querySubjectIdsWithRecords` ignored `ability` entirely. The outer subject query already constrained the result, so this was **not** a data leak — but the inner query read records the caller had no permission to read, and was inconsistent with every other query in the service.

## Tests

The three existing `hasRecord` tests are updated to the new call and still assert the same outcomes. Two added:

- the ids are grouped in the database, asserting `groupBy` is used and `findMany` is **not** — a guard against the `distinct` form coming back;
- the record query carries the caller's ability rather than being unconstrained.

## Checks

- `tsc --noEmit` on `apps/api` — clean
- `eslint src` — clean
- `prettier --check` — clean
- `vitest run` (whole workspace) — 254 passed, 1 skipped, 1 failed

The one failure is `instrument-records.service.spec.ts > upload > should create records with pending set`. This branch is cut from `main` so it carries that pre-existing failure; **#1422 fixes it**.

---

## Rebased onto `main`, review item addressed

Rebased onto `26ec89168`; CI green.

**Copilot: `expect(call.where.AND[0]).not.toStrictEqual({})` did not prove anything.** Correct, and
worse than it looks — an unconditional `read` rule legitimately produces `{}`, so that assertion would
have passed against a query that ignored the caller entirely. The test now builds a *conditioned*
ability and compares the clause against `accessibleQuery(ability, 'read', 'InstrumentRecord')`
directly, with `accessibleQuery` imported as suggested.

Verified by mutation: passing `undefined` in place of the ability now fails the test, where before it
passed.

---

## Round 2: rebased onto `5bfe4b2dc`, review items addressed

**1. The 500 — confirmed and fixed.** Reproduced exactly as you described: with a STANDARD ability,
`accessibleQuery(ability, 'read', 'InstrumentRecord')` throws
`ForbiddenError: It's not allowed to run "read" on "InstrumentRecord"` rather than returning `{}`.
`querySubjectIdsWithRecords` now returns an empty list for a caller holding no such rule, before
`accessibleQuery` is reached — the contract you settled on. The scoping itself stays.

**2. A unit test for that caller.** Added, asserting the empty list rather than a throw, with the
ability built through `AbilityFactory` with a STANDARD payload. Verified by mutation: removing the
guard fails it with that `ForbiddenError`.

**3. End-to-end.** `datahub.spec.ts` now toggles "With records only" and asserts the table empties,
plus an API-level case asserting a standard user gets an answer rather than a 500. Both verified by
mutation. This needed the `datahub-filter-has-records` and `datahub-filters-trigger` testids and
filter locators you flagged as missing.

The narrowing test has to know exactly what its group holds, and the `roleAccount` group is shared by
every spec in a worker, so this adds an `isolatedGroupManager` fixture. **#1425 introduces the same
fixture** — whichever merges second should drop its copy.

**4. The `not.toHaveBeenCalled()` assertion.** Dropped, agreed — the `groupBy` assertion above it
already rules out the `distinct` form.

The comment recording why the relation filter was rejected is kept, as you asked. Group-visibility
consequence stays as decided.

---

## Round 3: rebased onto `f97706ea8`, review items addressed

**1. The null-`groupId` note.** In the doc comment of `querySubjectIdsWithRecords`, next to the
STANDARD short-circuit, as you asked: the `accessibleQuery` clause drops records whose `groupId` is
null, because a group manager's rule is `groupId: { in: [...] }` and Prisma's `in` never matches
null.

**2. Rebase.** The app code did not change. Its patch is byte-identical to the reviewed one. Every
conflict was in `testing/`, which `main` reworked under this branch:

- `authenticateAs` on `main` now takes `$LoginCredentials`, and `appState` moved into the autouse
  `seedAppState` fixture. So `isolatedGroupManager` is now create group → create user →
  `authenticateAs(credentials)`, and the `injectAuth` helper this branch added is gone. The commit
  message that described that helper no longer mentions it.
- `datahub.spec.ts` keeps `main`'s three new tests, with this branch's two after them.
  `DatahubPage` keeps `main`'s `searchInput`. `api-client.ts` keeps `main`'s `findGroupById`,
  `setMailEnabled` and `updateUser` next to this branch's helpers.

`main` changed `ability.factory.ts` after the last review. The diff adds only an early return for a
user who must reset their password. It does not touch the STANDARD rules. The STANDARD unit test
and the e2e case that asserts `[]` both still pass. `apps/api/AGENTS.md` on `main` now also
recommends this exact short-circuit (`ability.can` before `accessibleQuery`).

`isolatedGroupManager`, its `testing/AGENTS.md` entry, `findInstrumentIdByName`, `uploadRecords`
and `HAPPINESS_RECORD` are now identical to #1425's copies. After this PR merges, #1425 conflicts
only where both PRs append to the same place (`datahub.spec.ts`, `DatahubPage`, the
`api-client.ts` imports).

**3. Commit scope.** `test(e2e)` is now `test(testing)`: `e2e` is not a workspace, and Commitlint
warns on it.

**Checks** at `8ab6ea2c7`:

- `pnpm lint`: 34/34 tasks, no files rewritten
- `pnpm test`: 142 files, 1160 tests passed
- `pnpm test:e2e`: 174 passed, 1 failed. The one failure is `subject-detail.spec.ts` › "should plot a selected measure on the graph tab", which no PR touches. It is flaky under full-suite load: "element was detached from the DOM" on the Measures button. It passes on its own, 3 of 3 runs.
- `commitlint --from origin/main`: clean

---

## Round 4: Copilot review of `8ab6ea2c7`

- **`as any` on the STANDARD token payload.** Fixed in `4eb3c58c9`. The payload is now complete and
  typed, with `mustResetPassword: false`. The test still fails when the guard is removed.
- **"a `instrumentRecords`".** Fixed in `b9b6493af`.

**Checks** at `b9b6493af`: `pnpm lint` 34/34 tasks, `pnpm test` 1160 passed. The e2e tests did not
change: this round changed only a unit test and a comment. CI green at this head.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_014ZEJLwa7jwD8sKzcE4PfSc
https://claude.ai/code/session_01KLfKsdHT1arfrc2fdT1fWq

### CODE REVIEW

**Comment by joshunrau** [comment]
CI is red on this branch, so I have not reviewed it yet.

**Failing check:** `lint-and-test` — [run 29771462892](https://github.com/DouglasNeuroInformatics/OpenDataCapture/actions/runs/29771462892/job/88450412133)

```
FAIL  api  src/instrument-records/__tests__/instrument-records.service.spec.ts
  > InstrumentRecordsService > upload > should create records with pending set,
    so they are not excluded from queries filtering on the field
AssertionError: expected "vi.fn()" to be called with arguments: [ { data: [ ObjectContaining{ pending: false } ] } ]
```

That test is not in your diff and no longer exists on `main` — it was removed in `722f0a0b0` ("test(api): update stale pending-on-create assertion for instrument records", 2026-07-20 22:39 EDT), which landed after this run. The result is stale.

Please merge or rebase onto current `main` and push to re-run CI. I'll pick the review back up once it's green.

**Comment by joshunrau** [comment]
**Suggested model: Sonnet 5**

Sizing the remaining work on this PR for whoever picks it up, human or agent. The branch is mergeable and the only blocker is stale CI, so remediation is merging `main` and pushing.

**Review by joshunrau — CHANGES_REQUESTED** [review]
Requesting changes so this tracks as needing author action.

**Blocking:** stale CI only — the branch is mergeable. Details in [my earlier comment](https://github.com/DouglasNeuroInformatics/OpenDataCapture/pull/1426#issuecomment-5094051773): the failing test isn't in your diff and was removed from `main` in `722f0a0b0`, after this run was created. Merge `main` and push to get a fresh result.

No content review has been done on this branch yet — it's gated on CI being green.

**Review by joshunrau — CHANGES_REQUESTED** [review]
The perf work is right and the evidence for it is good — `groupBy` genuinely pushes the distinct into
MongoDB where `distinct` did not, and the comment recording why the `instrumentRecords: { some: ... }`
relation filter was rejected is worth more than the code change itself. Please keep it. The
`accessibleQuery` pin in the new test is also a real improvement over the `not.toStrictEqual({})`
version.

One blocking problem, in the other half of the change:

1. `accessibleQuery(ability, 'read', 'InstrumentRecord')` at `apps/api/src/subjects/subjects.service.ts:178`
   **throws** when the caller holds no `read` rule for `InstrumentRecord`. `@casl/prisma`'s
   `accessibleBy` proxy raises `ForbiddenError` rather than returning `{}` in that case, and every
   STANDARD user is in that case — `ability.factory.ts` grants them `create` on `InstrumentRecord` but
   never `read`. This route's guard is `@RouteAccess({ action: 'read', subject: 'Subject' })`, which
   STANDARD users pass, so they reach the service and the CASL error escapes as a **500** (it is not
   an `HttpException`, so libnest's filter maps it to Internal Server Error). On `main` the same
   request returns a correct 200. I reproduced this by driving `SubjectsService.find({ hasRecord: true })`
   with an ability built from `AbilityFactory.createForPayload({ basePermissionLevel: 'STANDARD', ... })`.

   Nobody hits it through the web UI — `useNavItems.ts:63` hides the datahub entirely without
   `read InstrumentRecord` — but `GET /v1/subjects?hasRecord=true` is a documented endpoint and the
   guard admits these callers. Every other `accessibleQuery(..., 'InstrumentRecord')` in the codebase
   sits behind a route guard naming the same subject, which is why this shape has not bitten before.

   Joshua has settled the contract: such a caller should get an **empty list** — the honest reading
   of "subjects whose records you can see" — not a 403 and not the pre-PR behaviour of ignoring
   record permissions here. So the no-rule case needs handling before `accessibleQuery` is reached,
   or the `ForbiddenError` catching, with an empty result returned either way.

2. Please add a unit test for that caller, asserting the empty list rather than a throw. The new
   ability test builds a permission set that can read `InstrumentRecord`, so it exercises the
   conditioned branch but never the no-rule branch — which is exactly the one that breaks.

3. There is no e2e test, which the root `CLAUDE.md` asks for. `testing/src/specs/datahub.spec.ts`
   only checks that the page header renders. Toggling "With records only" and asserting the table
   narrows would cover both the perf path and the permissions change. The checkbox at
   `apps/web/src/routes/_app/datahub/index.tsx:169` has no `data-testid` and
   `testing/src/pages/_app/datahub/index.page.ts` exposes no filter locator, so both need adding.

4. Minor: `expect(prismaClient.instrumentRecord.findMany).not.toHaveBeenCalled()` in
   `subjects.service.spec.ts:105` asserts an implementation detail. The `groupBy` assertion just above
   it already stops the `distinct` form coming back, so the negative one mostly adds a way for an
   unrelated future refactor to fail the test.

On the other consequence of the scoping — a group manager no longer seeing subjects whose records
live only in groups they cannot read — Joshua has confirmed that is intended, as a disclosure fix
consistent with how `/v1/instrument-records` already scopes. Nothing to change there.

If you hand this to Claude Code, **Fable 5** is the right size — it is auth and group-scoping work,
and the fix has to be carried consistently through the service, its unit tests and a new e2e spec.

Reviewed at commit 53c8db8b1c966c2cc4e7788dc321a6771726cd98.

**Review by joshunrau — CHANGES_REQUESTED** [review]
Thanks — everything from the last round landed, lint is clean across all 33 workspaces here and the api
suite passes 156/156. The scoping question this raised has been settled in favour of what you wrote:
ungrouped (admin-created) records should **not** count toward "with records only" for a group manager,
so no change to the filter is wanted.

Worth stating plainly, because it makes this more than a perf change: on `main` this subquery ran with
no ability filter at all (`where: groupId ? { groupId } : {}`), so a group manager's "has records"
answer was computed from every record in the collection. Adding `accessibleQuery` here fixes that.

One thing to add before merge:

1. `apps/api/src/subjects/subjects.service.ts:182-187` — please note in the doc comment that the
   `accessibleQuery` clause drops records with a null `groupId`, since a group manager's rule matches
   `groupId: { in: groupIds }` and Prisma's `in` does not match null. It is the behaviour we want, but
   it is invisible at the call site and it is precisely why admin-created records disappear from a
   manager's filtered view. The block already explains the STANDARD short-circuit just above, so this
   is the same pattern in the same place.

No change needed for STANDARD users, by the way — I checked, and both datahub entries in
`useNavItems.ts` are already gated on `read InstrumentRecord`, so they have no UI path to the toggle.

On ordering: this one is planned to merge first, ahead of #1425 and #1427. Nothing for you to do about
the shared `isolatedGroupManager` fixture — the later PRs will drop their copies.

If you hand this to Claude Code, **Sonnet 5** is the right size — a single comment in one file,
mirroring the explanatory comment already three lines above it.

Reviewed at commit 3bd9380fa2f5c9ad51029ddf47fcee24e6239abe.

**Review by joshunrau — CHANGES_REQUESTED** [review]
This branch no longer merges cleanly into `main` — GitHub reports it as conflicting, so it can't be reviewed or merged as it stands. `main` has moved since the last pass (#1444 and the e2e-test work landed).

Please merge or rebase onto current `main` and push. Review picks up again once the branch is mergeable and CI is green — the new commits since the last review haven't been looked at yet.

Reviewed at commit 3a73274225383fbd9b2b95fdf1f21885966255b8.

File: apps/api/src/subjects/__tests__/subjects.service.spec.ts [hunk 1]
```diff
@@ -98,18 +94,60 @@ describe('SubjectsService', () => {
...
+    // A STANDARD user holds `create` but not `read` on InstrumentRecord, and this route's guard names
+    // `read Subject`, so they reach the service. `accessibleQuery` throws on an ability with no rule
+    // for the subject at all, which escapes as a 500 rather than the empty list they should see.
+    it('should return an empty list for a caller who may read no records, rather than throwing', async () => {
+      const abilityFactory = new AbilityFactory(MockFactory.createMock(LoggingService) as unknown as LoggingService);
+      const ability = abilityFactory.createForPayload({
+        basePermissionLevel: 'STANDARD',
+        groups: [{ id: 'group-1' }],
+        id: 'user-1'
+      } as any);
```
[the remark this answers is not in the document]
Comment (gdevenyi) [reply apps/api/src/subjects/__tests__/subjects.service.spec.ts]: Fixed in 4eb3c58c9. The payload is now a complete `Omit<TokenPayload, 'permissions'>` with `mustResetPassword: false`, and there is no cast. `groups` is empty, so no `Group` fixture is needed: the missing `read InstrumentRecord` rule does not depend on groups. With the guard removed, the test still fails with the `ForbiddenError`.

File: apps/api/src/subjects/subjects.service.ts [hunk 2]
```diff
@@ -158,12 +159,36 @@ export class SubjectsService {
...
-      select: { subjectId: true },
-      where: groupId ? { groupId } : {}
+  /**
+   * The ids of subjects having at least one record, for the "with records only" filter.
+   *
+   * Grouped by the database rather than deduplicated after the fact: `distinct` is applied by the
+   * prisma query engine, so it returns one row per *record* over the wire and collapses them only
+   * once they have arrived.
+   *
+   * Deliberately not expressed as a `instrumentRecords: { some: ... }` relation filter on Subject.
```
[the remark this answers is not in the document]
Comment (gdevenyi) [reply apps/api/src/subjects/subjects.service.ts]: Fixed in b9b6493af.

**Comment by gdevenyi** [comment]
## Rebased onto `fe0985034`

The app code did not change. Two conflicts, both in `testing/`, both from `main` adding next to this branch:

- `testing/AGENTS.md` fixtures table: kept `main`'s `setUserPermissions()` note and this branch's `isolatedGroupManager` row.
- `api-client.ts`: kept `main`'s new `findUserById` and this branch's `findInstrumentIdByName`, in the order eslint sorts them.

`main`'s group-scoped grants (`8c4cbeeed`) do not change the STANDARD short-circuit: `ability.can('read', 'InstrumentRecord')` is still false for a caller with no rule, and true for a scoped grant, which then goes through `accessibleQuery` as before.

**Checks:** `pnpm lint` clean. `pnpm test`: all pass except 2 `react-core` InstrumentRenderer tests (`localStorage.getItem` undefined), which fail the same way on `main`. `pnpm test:e2e`: 200/200.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_015wLMGmHorLgMr7dzbbXjMF
